src/m3ninx/index/segment/fst/encoding/docs/compressed_index.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package docs
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3/src/m3ninx/index"
+	"github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding"
+	"github.com/m3db/m3/src/m3ninx/postings"
+)
+
+// compressedIndexWriter is a writer for the index file for documents.
+type compressedIndexWriter struct {
+	writer io.Writer
+	enc    *encoding.Encoder
+	ready  bool
+	prev   postings.ID
+}
+
+// newCompressedIndexWriter returns a new compressedIndexWriter.
+func newCompressedIndexWriter(w io.Writer) indexWriter {
+	iw := &compressedIndexWriter{
+		writer: w,
+		enc:    encoding.NewEncoder(initialIndexEncoderLen),
+	}
+	return iw
+}
+
+// Write writes the offset for an id. IDs must be written in increasing order but can be
+// non-contiguous.
+func (w *compressedIndexWriter) Write(id postings.ID, o offset) error {
+	if !w.ready {
+		w.writeMetadata(id)
+		w.ready = true
+	} else {
+		if id <= w.prev {
+			return fmt.Errorf("postings IDs must be monotonically increasing: received %v but previous ID was %v", id, w.prev)
+		}
+		for i := 0; i < int(id-w.prev)-1; i++ {
+			w.enc.PutUint64(emptyID)
+			w.enc.PutUint64(emptyID)
+		}
+	}
+
+	w.enc.PutUint64(o.pageOffset)
+	w.enc.PutUint64(o.docOffset)
+	w.prev = id
+
+	return w.write()
+}
+
+func (w *compressedIndexWriter) writeMetadata(id postings.ID) {
+	w.enc.PutUint64(uint64(id))
+}
+
+func (w *compressedIndexWriter) write() error {
+	b := w.enc.Bytes()
```

Comment:
Is there a way to avoid having to do this? Like have the encoder write directly into the writer (or a buffered writer)